File: src/pick_and_place_pkg/pick_and_place_pkg/pick_and_place.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import Float64MultiArray
from moveit.planning import MoveItPy, PlanningComponent

class PickAndPlace(Node):
    def __init__(self):
        super().__init__('pick_and_place_node')

        # ▶ 파라미터 파일 없이 생성
        self.moveit_py = MoveItPy(
            node_name='pick_and_place',            # rclpy 노드 이름
            provide_planning_service=True          # planning 서비스 제공
        )
        self.arm: PlanningComponent = self.moveit_py.get_planning_component('ur_manipulator')
        self.get_logger().info('MoveItPy initialized (parameters from launch)')

        # 그리퍼 제어 퍼블리셔
        self.gripper_pub = self.create_publisher(
            Float64MultiArray,
            '/forward_position_gripper_controller/commands',
            10
        )

    def go_to_pose(self, pose: PoseStamped) -> bool:
        self.arm.set_start_state_to_current_state()
        self.arm.set_goal_state(pose_stamped_msg=pose, pose_link='ee_link')
        plan = self.arm.plan()
        if plan.joint_trajectory.points:
            self.get_logger().info('Plan found; executing...')
            self.arm.execute(plan.joint_trajectory)
            return True
        else:
            self.get_logger().error('Planning failed!')
            return False

    def control_gripper(self, position: float):
        msg = Float64MultiArray(data=[position])
        self.gripper_pub.publish(msg)
        self.get_logger().info(f'Gripper cmd: {position}')

    def run(self):
        # --- PICK ---
        pick = PoseStamped()
        pick.header.frame_id = 'base_link'
        pick.pose.position.x = 0.5
        pick.pose.position.y = 0.0
        pick.pose.position.z = 0.2
        pick.pose.orientation.w = 1.0

        if not self.go_to_pose(pick):
            return
        self.control_gripper(0.0)  # close
        self.get_logger().info('Picked, waiting 1s...')
        self.get_clock().sleep_for(rclpy.duration.Duration(seconds=1))

        # --- PLACE ---
        place = PoseStamped()
        place.header.frame_id = 'base_link'
        place.pose.position.x = 0.3
        place.pose.position.y = -0.3
        place.pose.position.z = 0.2
        place.pose.orientation.w = 1.0

        if not self.go_to_pose(place):
            return
        self.control_gripper(0.8)  # open
        self.get_logger().info('Placed, done.')

def main(args=None):
    rclpy.init(args=args)
    node = PickAndPlace()
    node.run()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/move_test_cpp/src/pick_and_place.cpp
#include <rclcpp/rclcpp.hpp>
#include <moveit/move_group_interface/move_group_interface.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <std_msgs/msg/float64_multi_array.hpp>
#include <chrono>

#include <tf2/LinearMath/Quaternion.h>


using namespace std::chrono_literals;

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  auto node = rclcpp::Node::make_shared("pick_and_place_cpp");

  // MoveGroupInterface for UR5e manipulator
  moveit::planning_interface::MoveGroupInterface move_group(node, "ur1_manipulator");
  move_group.setPlanningTime(10.0);
  move_group.setStartStateToCurrentState();
  move_group.setGoalPositionTolerance(0.01);   // 1cm
  move_group.setGoalOrientationTolerance(0.05); // 약 3°
  move_group.allowReplanning(true);
  // 여러 번 시도
  move_group.setNumPlanningAttempts(5);
  move_group.setMaxVelocityScalingFactor(0.5);   // 속도의 50%
  move_group.setMaxAccelerationScalingFactor(0.5);

  // Gripper command publisher
  auto gripper_pub = node->create_publisher<std_msgs::msg::Float64MultiArray>(
    "/ur1/forward_position_gripper_controller/commands", 10);

  // Open gripper
  std_msgs::msg::Float64MultiArray open_cmd;
  open_cmd.data = {0.1};  // adjust value for open position
  gripper_pub->publish(open_cmd);
  RCLCPP_INFO(node->get_logger(), "Gripper opened");


  // --- PICK ---
  geometry_msgs::msg::PoseStamped pick_pose;
  pick_pose.header.frame_id = "ur1_base_link";
  
  // 위치 설정
  pick_pose.pose.position.x = 0.5;
  pick_pose.pose.position.y = 0.0;
  pick_pose.pose.position.z = 0.2;
  
  // tf2 쿼터니언 생성
  tf2::Quaternion q;
  q.setRPY(M_PI, 0.0, 0.0);
  // geometry_msgs::msg::Quaternion 으로 직접 복사
  pick_pose.pose.orientation.x = q.x();
  pick_pose.pose.orientation.y = q.y();
  pick_pose.pose.orientation.z = q.z();
  pick_pose.pose.orientation.w = q.w();
  
  // 플래닝
  move_group.setStartStateToCurrentState();
  move_group.setPoseTarget(pick_pose);

  moveit::planning_interface::MoveGroupInterface::Plan pick_plan;
  if (move_group.plan(pick_plan) == moveit::core::MoveItErrorCode::SUCCESS 
      && !pick_plan.trajectory.joint_trajectory.points.empty()) {
    RCLCPP_INFO(node->get_logger(), "Pick plan found, executing");
    move_group.execute(pick_plan);
  } else {
    RCLCPP_ERROR(node->get_logger(), "Pick planning failed");
    return 1;
  }

  // Close gripper
  std_msgs::msg::Float64MultiArray close_cmd;
  close_cmd.data = {0.7};  // adjust value for closed position
  gripper_pub->publish(close_cmd);
  RCLCPP_INFO(node->get_logger(), "Gripper closed");
  rclcpp::sleep_for(1s);

  // --- PLACE ---
  geometry_msgs::msg::PoseStamped place_pose;
  place_pose.header.frame_id = "ur1_base_link";
  place_pose.pose.position.x = 0.3;
  place_pose.pose.position.y = -0.3;
  place_pose.pose.position.z = 0.0;

  place_pose.pose.orientation.x = q.x();
  place_pose.pose.orientation.y = q.y();
  place_pose.pose.orientation.z = q.z();
  place_pose.pose.orientation.w = q.w();

  move_group.setStartStateToCurrentState();
  move_group.setPoseTarget(place_pose);
  moveit::planning_interface::MoveGroupInterface::Plan place_plan;
  if (move_group.plan(place_plan) == moveit::core::MoveItErrorCode::SUCCESS 
      && !place_plan.trajectory.joint_trajectory.points.empty()) {
    RCLCPP_INFO(node->get_logger(), "Place plan found, executing");
    move_group.execute(place_plan);
  } else {
    RCLCPP_ERROR(node->get_logger(), "Place planning failed");
    return 1;
  }

  // Open gripper
  open_cmd.data = {0.1};  // adjust value for open position
  gripper_pub->publish(open_cmd);
  RCLCPP_INFO(node->get_logger(), "Gripper opened");

  rclcpp::shutdown();
  return 0;
}
